crates/iceberg/src/io.rs

```diff
@@ -0,0 +1,451 @@
+    ///
+    /// If this is `None`, then [`FileIOBuilder::build`](FileIOBuilder::build) will build a local file io.
+    scheme_str: Option<String>,
+    /// Arguments for operator.
+    props: HashMap<String, String>,
+}
+
+impl FileIOBuilder {
+    /// Creates a new builder with scheme.
+    pub fn new(scheme_str: impl ToString) -> Self {
+        Self {
+            scheme_str: Some(scheme_str.to_string()),
+            props: HashMap::default(),
+        }
+    }
+
+    /// Creates a new builder for local file io.
+    pub fn new_local_file_io() -> Self {
+        Self {
+            scheme_str: None,
+            props: HashMap::default(),
+        }
+    }
+
+    /// Add argument for operator.
+    pub fn with_prop(mut self, key: impl ToString, value: impl ToString) -> Self {
+        self.props.insert(key.to_string(), value.to_string());
+        self
+    }
+
+    /// Add argument for operator.
+    pub fn with_props(
+        mut self,
+        args: impl IntoIterator<Item = (impl ToString, impl ToString)>,
+    ) -> Self {
+        self.props
+            .extend(args.into_iter().map(|e| (e.0.to_string(), e.1.to_string())));
+        self
+    }
+
+    /// Builds [`FileIO`].
+    pub fn build(self) -> Result<FileIO> {
+        let storage = Storage::build(self)?;
+        Ok(FileIO {
+            inner: Arc::new(storage),
+        })
+    }
+}
+
+impl FileIO {
+    /// Deletes file.
+    pub async fn delete(&self, path: impl AsRef<str>) -> Result<()> {
+        let (op, relative_path) = self.inner.create_operator(&path)?;
+        Ok(op.delete(relative_path).await?)
+    }
+
+    /// Check file exists.
+    pub async fn is_exist(&self, path: impl AsRef<str>) -> Result<bool> {
+        let (op, relative_path) = self.inner.create_operator(&path)?;
+        Ok(op.is_exist(relative_path).await?)
+    }
+
+    /// Creates input file.
+    pub fn new_input(&self, path: impl AsRef<str>) -> Result<InputFile> {
+        let (op, relative_path) = self.inner.create_operator(&path)?;
+        let path = path.as_ref().to_string();
+        let relative_path_pos = path.len() - relative_path.len();
+        Ok(InputFile {
+            op,
+            path,
+            relative_path_pos,
+        })
+    }
+
+    /// Creates output file.
+    pub fn new_output(&self, path: impl AsRef<str>) -> Result<OutputFile> {
+        let (op, relative_path) = self.inner.create_operator(&path)?;
+        let path = path.as_ref().to_string();
+        let relative_path_pos = path.len() - relative_path.len();
+        Ok(OutputFile {
+            op,
+            path,
+            relative_path_pos,
+        })
+    }
+}
+
+/// Input file implementation.
+#[derive(Debug)]
+pub struct InputFile {
+    op: Operator,
+    // Absolution path of file.
+    path: String,
+    // Relative path of file to uri, starts at [`relative_path_pos`]
+    relative_path_pos: usize,
+}
+
+/// Input stream for reading.
+pub trait InputStream: AsyncRead + AsyncSeek {}
```

Comment:
How about defining our own traits, such as `FileRead` and `FileWrite`, to make our API more consistent?

```diff
- pub trait InputStream: AsyncRead + AsyncSeek {}
- pub async fn reader(&self) -> Result<impl InputStream> {}
- pub async fn writer(&self) -> Result<impl AsyncWrite> {}

+ pub trait FileRead: AsyncRead + AsyncSeek {}
+ pub async fn reader(&self) -> Result<impl FileRead> {}
+ pub trait FileWrite: AsyncWrite {}
+ pub async fn writer(&self) -> Result<impl FileWrite> {}
```

Reply:
Personally, I have no preference for the naming convention. I choose to follow java/python implementation's naming since it make review and feature parity check easier.

Reply:
The `FileRead`/`FileWrite` trait do look better and more consistent.

Reply:
While the traits aliases make the implementation more consistent, I'm not sure whether they convey the same information when looking at the function signature in the documentation. The traits `AsyncRead` and `AsyncWrite` are often used traits in the rust ecosystem and users have an understanding of the concepts. 

I remember that I was looking for an object store implementation and I was explicitly searching for `AsyncWrite` because the library I was using required a type that implements `AsyncWrite`. I'm not sure if that is obvious with `impl FileWrite`.

Just something to think about.